crates/wasi/src/preview2/host/filesystem/sync.rs

```diff
@@ -316,97 +398,167 @@ impl<T: async_filesystem::Host> sync_filesystem::Host for T {
+                self,
+                Resource::new_borrow(fd.rep()),
+            )
+            .await
         })?)
     }
 
-    fn unlock(&mut self, fd: sync_filesystem::Descriptor) -> Result<(), sync_filesystem::Error> {
+    fn unlock(
+        &mut self,
+        fd: Resource<sync_filesystem::Descriptor>,
+    ) -> Result<(), sync_filesystem::Error> {
         Ok(in_tokio(async {
-            async_filesystem::Host::unlock(self, fd).await
+            async_filesystem::HostDescriptor::unlock(self, Resource::new_borrow(fd.rep())).await
         })?)
     }
 
     fn read_via_stream(
         &mut self,
-        fd: sync_filesystem::Descriptor,
+        fd: Resource<sync_filesystem::Descriptor>,
         offset: sync_filesystem::Filesize,
-    ) -> Result<streams::InputStream, sync_filesystem::Error> {
-        Ok(async_filesystem::Host::read_via_stream(self, fd, offset)?)
+    ) -> Result<Resource<streams::InputStream>, sync_filesystem::Error> {
+        Ok(Resource::new_own(
+            async_filesystem::HostDescriptor::read_via_stream(
+                self,
+                Resource::new_borrow(fd.rep()),
+                offset,
+            )?
+            .rep(),
+        ))
     }
 
     fn write_via_stream(
         &mut self,
-        fd: sync_filesystem::Descriptor,
+        fd: Resource<sync_filesystem::Descriptor>,
         offset: sync_filesystem::Filesize,
-    ) -> Result<streams::OutputStream, sync_filesystem::Error> {
-        Ok(async_filesystem::Host::write_via_stream(self, fd, offset)?)
+    ) -> Result<Resource<streams::OutputStream>, sync_filesystem::Error> {
+        Ok(Resource::new_own(
+            async_filesystem::HostDescriptor::write_via_stream(
+                self,
+                Resource::new_borrow(fd.rep()),
+                offset,
+            )?
+            .rep(),
+        ))
     }
 
     fn append_via_stream(
         &mut self,
-        fd: sync_filesystem::Descriptor,
-    ) -> Result<streams::OutputStream, sync_filesystem::Error> {
-        Ok(async_filesystem::Host::append_via_stream(self, fd)?)
+        fd: Resource<sync_filesystem::Descriptor>,
+    ) -> Result<Resource<streams::OutputStream>, sync_filesystem::Error> {
+        Ok(Resource::new_own(
+            async_filesystem::HostDescriptor::append_via_stream(
+                self,
+                Resource::new_borrow(fd.rep()),
+            )?
+            .rep(),
+        ))
     }
 
     fn is_same_object(
         &mut self,
-        a: sync_filesystem::Descriptor,
-        b: sync_filesystem::Descriptor,
+        a: Resource<sync_filesystem::Descriptor>,
+        b: Resource<sync_filesystem::Descriptor>,
     ) -> anyhow::Result<bool> {
         Ok(in_tokio(async {
-            async_filesystem::Host::is_same_object(self, a, b).await
+            async_filesystem::HostDescriptor::is_same_object(
+                self,
+                Resource::new_borrow(a.rep()),
+                Resource::new_borrow(b.rep()),
```

Comment:
Bah this is unfortunate (as are all the other instances of this in this file). If I understand this correctly, this is because the bindings generated. have two distinct `Descriptor` types meaning that they're not compatible?

If so I can try to add a feature to `bindgen!` so resources can be specified as opposed to always-generated

Reply:
Yes, exactly. All the `sync` adapter code has to jump through these hoops for that reason.